### PR TITLE
[4.0] Namespace com_fields

### PR DESCRIPTION
**Pull request by laoneo**
Namespace com_fields.

Some extras:
- The libraries classes are now in a plugin namespace.
- The model can detect the factory by itself when it is namespaced, doesn't default to the LegacyFactory only
- Introduced a new classmap file for the core extensions, where we can put in alias which we need and are extension specific
- Models are used directly and not trough `JModelLegacy::getInstance()`
- Removed some invalid code from merges

### CODE REVIEW

File: administrator/components/com_fields/Controller/Field.php [hunk 1]
```diff
@@ -31,18 +36,23 @@ class FieldsControllerField extends JControllerForm
...
-	 * @param   array  $config  A named array of configuration variables.
+	 * @param   array                $config   An optional associative array of configuration settings.
+	 * Recognized key values include 'name', 'default_task', 'model_path', and
+	 * 'view_path' (this list is not meant to be comprehensive).
+	 * @param   MvcFactoryInterface  $factory  The factory.
+	 * @param   CmsApplication       $app      The JApplication for the dispatcher
+	 * @param   \JInput              $input    Input
 	 *
-	 * @since   3.7.0
+	 * @since   3.0
```
Comment (brianteeman) [inline administrator/components/com_fields/Controller/Field.php]: did you mean to change this?

File: administrator/components/com_fields/Controller/Fields.php [hunk 2]
```diff
@@ -27,15 +31,15 @@ class FieldsControllerFields extends JControllerAdmin
...
-	 * @param   array   $config  The array of possible config values. Optional.
+	 * @param   string  $name    The name of the model.
+	 * @param   string  $prefix  The prefix for the PHP class name.
+	 * @param   array   $config  Array of configuration parameters.
 	 *
-	 * @return  FieldsModelField|boolean
+	 * @return  \Joomla\CMS\Model\Model
 	 *
-	 * @since   3.7.0
+	 * @since   1.6
```
Comment (brianteeman) [inline administrator/components/com_fields/Controller/Fields.php]: did you meant to change this

File: administrator/components/com_fields/Controller/Group.php [hunk 3]
```diff
@@ -35,17 +40,22 @@ class FieldsControllerGroup extends JControllerForm
...
-	 * @param   array  $config  A named array of configuration variables.
+	 * @param   array                $config   An optional associative array of configuration settings.
+	 * Recognized key values include 'name', 'default_task', 'model_path', and
+	 * 'view_path' (this list is not meant to be comprehensive).
+	 * @param   MvcFactoryInterface  $factory  The factory.
+	 * @param   CmsApplication       $app      The JApplication for the dispatcher
+	 * @param   \JInput              $input    Input
 	 *
-	 * @since   3.7.0
+	 * @since   3.0
```
Comment (brianteeman) [inline administrator/components/com_fields/Controller/Group.php]: again looks like you shouldnt have changed this

File: administrator/components/com_fields/Controller/Groups.php [hunk 4]
```diff
@@ -27,15 +31,15 @@ class FieldsControllerGroups extends JControllerAdmin
...
-	 * @param   array   $config  The array of possible config values. Optional.
+	 * @param   string  $name    The name of the model.
+	 * @param   string  $prefix  The prefix for the PHP class name.
+	 * @param   array   $config  Array of configuration parameters.
 	 *
-	 * @return  JModelLegacy|boolean  Model object on success; otherwise false on failure.
+	 * @return  \Joomla\CMS\Model\Model
 	 *
-	 * @since   3.7.0
+	 * @since   1.6
```
Comment (brianteeman) [inline administrator/components/com_fields/Controller/Groups.php]: 3.7.0

File: administrator/components/com_fields/Model/Fields.php [hunk 5]
```diff
@@ -16,17 +20,18 @@
...
-	 * Constructor.
+	 * Constructor
 	 *
-	 * @param   array  $config  An optional associative array of configuration settings.
+	 * @param   array                $config   An array of configuration options (name, state, dbo, table_path, ignore_request).
+	 * @param   MvcFactoryInterface  $factory  The factory.
 	 *
-	 * @see     JModelLegacy
-	 * @since   3.7.0
+	 * @since   1.6
```
Comment (brianteeman) [inline administrator/components/com_fields/Model/Fields.php]: 3.7.0

**Comment by brianteeman** [comment]
Please review all the version number @since changes - looks like a lot of copy paste errors. I started to comment inline on the code but there were too many

**Comment by laoneo** [comment]
Thanks @brianteeman, fixed the since params.

File: administrator/components/com_fields/Controller/Field.php [hunk 6]
```diff
@@ -106,12 +116,12 @@ protected function allowAdd($data = array())
 	 *
 	 * @return  boolean
 	 *
-	 * @since   1.6
+	 * @since   3.7.0
```
Comment (brianteeman) [inline administrator/components/com_fields/Controller/Field.php]: shouldnt this be since 1.6

Comment (brianteeman) [inline administrator/components/com_fields/Controller/Field.php]: or maybe it was  wrong before

Comment (laoneo) [inline administrator/components/com_fields/Controller/Field.php]: it was wrong before, correctly it should be done against staging.

Comment (brianteeman) [inline administrator/components/com_fields/Controller/Field.php]: ok

**Comment by brianteeman** [comment]
(title updated to show its for v4)

**Pull request merged**